design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+
+  /**
+   * Drop table. The underlying Kafka topic will not be deleted.
+   */
+  CompletableFuture<DropSourceResponse> dropTable(String tableName);
+
+  /**
+   * Drop table. The underlying Kafka topic may optionally be deleted.
+   */
+  CompletableFuture<DropSourceResponse> dropTable(String tableName, boolean deleteTopic);
+```
+where `DropSourceResponse` is actually the same as `DdlResponse`/`DmlResponse`/`CreateSourceResponse`/`InsertIntoResponse` above.
+
+Again, the implementations of `dropStream(...)` and `dropTable(...)` would be the same so we could instead have a single `dropSource(...)`, but the naming might be confusing.
+
+If we choose to keep them separate, there's an open question of whether the client should validate that `dropStream(...)` is not used to drop a table, and vice versa.
+IMO such validation would be introducing complexity without much benefit, though this point of ambiguity makes me prefer a single `dropSource(...)` if we can agree on a method name that's not confusing.
+
+Note that users can also execute `DROP <STREAM/TABLE>` requests via `executeDdl(...)` or `executDml(...)` above.
+
+#### `TERMINATE <queryId>`
+
+```
+  CompletableFuture<TerminateQueryResponse> terminatePersistentQuery(String queryId);
+```
+where `TerminateQueryResponse` is again the same as `DdlResponse`/`DmlResponse`/`CreateSourceResponse`/`InsertIntoResponse`/`DropSourceResponse` above.
+
+The method name `terminatePersistenQuery(...)` is to distinguish from `terminatePushQuery(...)` below.
+
+#### Connectors
+
+```
+  CompletableFuture<ConnectorInfo> createSourceConnector(String name, Map<String, String> properties);
+
+  CompletableFuture<ConnectorInfo> createSinkConnector(String name, Map<String, String> properties);
+
+  CompletableFuture<Void> dropConnector(String name);
+```
+where `ConnectorInfo` is from an Apache Kafka module ([link](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/ConnectorInfo.java)).
+
+Or would we rather not have a dependency on Apache Kafka in the ksqlDB client interfaces?
+
+### Admin operations
+
+#### `SHOW TOPICS`
+
+```
+  CompletableFuture<List<TopicInfo>> listTopics();
+```
+with
+```
+public interface TopicInfo {
+
+  String getName();
+
+  int getPartitions();
+  
+  List<Integer> getReplicasPerPartition();
+
+}
+```
+
+#### `SHOW <STREAMS/TABLES>`
+
+```
+  CompletableFuture<List<StreamInfo>> listStreams();
+
+  CompletableFuture<List<TableInfo>> listTables();
+```
+with
+```
+public interface StreamInfo {
+
+  String getName();
+
+  String getTopic();
+
+  String getFormat();
+  
+}
+```
+and
+```
+public interface TableInfo {
+
+  String getName();
+
+  String getTopic();
+
+  String getFormat();
+
+  boolean isWindowed();
+  
+}
+```
+
+I'm not sure whether it makes more sense for `StreamInfo#getFormat()` and `TableInfo#getFormat()` to return a string or an enum value.
```

Comment:
For this case, I think a String makes sense - especially if we ever want to support pluggable schemas (which is what schema registry has moved toward and they standardized on strings as the names for formats)